src/itwinai/torch/monitoring/monitoring.py

```diff
@@ -0,0 +1,191 @@
+    a multiprocessing.Manager object. Logs utilization into `log_dict` until
+    stop_flag.value is set to True.
+
+    Args:
+        node_idx: The index of the compute node that the function is called by, used
+            for logging purposes.
+        num_local_gpus: Number of GPUs on the current compute node.
+        num_global_gpus: Number of GPUs on all nodes combined.
+        strategy: Which distributed strategy is being used, e.g. "ddp" or "horovod".
+        log_dict: Dictionary for storing logging data on. Should be managed by a
+            multiprocessing.Manager object.
+        stop_flag: Shared value telling the function when to stop logging. Should be
+            managed by a multiprocessing.Manager object.
+        probing_interval: How long to wait between each time a read of the GPU
+            utilization is done.
+        warmup_time: How long to wait before logging starts, allowing the training to
+            properly start before reading.
+
+    """
+
+    if not set(logging_columns).issubset(set(log_dict.keys())):
+        missing_columns = set(logging_columns) - set(log_dict.keys())
+        raise ValueError(
+            f"log_dict is missing the following columns: {missing_columns}"
+        )
+
+    nvmlInit()
+    time.sleep(warmup_time)
+
+    sample_idx = 0
+    while not stop_flag.value:
+        for idx in range(num_local_gpus):
+            handle = nvmlDeviceGetHandleByIndex(idx)
+            utilization_rates = pynvml.nvmlDeviceGetUtilizationRates(handle)
+
+            gpu_util = utilization_rates.gpu
+            power = pynvml.nvmlDeviceGetPowerUsage(handle)
+            power = power / 1000  # mW -> W
+
+            log_dict["sample_idx"].append(sample_idx)
+            log_dict["utilization"].append(gpu_util)
+            log_dict["power"].append(power)
+            log_dict["local_rank"].append(idx)
+            log_dict["node_idx"].append(node_idx)
+            log_dict["num_global_gpus"].append(num_global_gpus)
+            log_dict["strategy"].append(strategy_name)
+            log_dict["probing_interval"].append(probing_interval)
+
+        sample_idx += 1
+
+        time.sleep(probing_interval)
+
+
+def measure_gpu_utilization(method: Callable) -> Callable:
+    """Decorator for measuring GPU utilization and storing it to a .csv file."""
+
+    def write_logs_to_file(utilization_logs: List[Dict], output_path: Path) -> None:
+        dataframes = []
+        for log in utilization_logs:
+            if len(log) == 0:
+                continue
+            dataframes.append(pd.DataFrame(log))
+
+        log_df = pd.concat(dataframes)
+        log_df.to_csv(output_path, index=False)
+        print(f"Writing DataFrame to '{output_path}'.")
+
+    @functools.wraps(method)
+    def measured_method(self: TorchTrainer, *args, **kwargs) -> Any:
+        gpu_probing_interval = 1
+        warmup_time = 5
+
+        strategy = self.strategy
+        strategy.init()
```

Comment:
Move strategy creation and destroy outside of the decorators, as it should not be their responsibility to manage that. In addition, it would make debugging more complicated.